### DISCUSSION TITLE
Learning methods

### DISCUSSION THREAD

**Discussion by RasmusBrostroem** [RasmusBrostroem/ConnectFourRL#43] [Ideas]
This discussion can be used to share ideas on how the learning can happen and which methods in RL we could use to get better agents.

**Comment by RasmusBrostroem** [comment]
Instead of using policy gradient we can maybe take some ideas from this method: https://openai.com/blog/openai-baselines-ppo/

**Reply by jbirkesteen** [reply]
Looks promising in the blog post. Let's discuss it later today and keep it in mind Saturday!

**Reply by jbirkesteen** [reply]
Just stumbled across this video of an [AI learning a cube to escape rooms with obstructions](https://youtu.be/2tamH76Tjvw), which uses PPO and a relatively simple neural network. *See the poster's pinned comment for details*.

**Comment by jbirkesteen** [comment]
Deepmind published a new method the other day, which was able to beat expert human players in the game of Stratego!
Their method is shortly presented in this [blog post](https://www.deepmind.com/blog/mastering-stratego-the-classic-game-of-imperfect-information), but the [paper](https://www.science.org/doi/10.1126/science.add4679) unfortunately seems to be behind a paywall. However, the code base is publicly available as [open source](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/algorithms/rnad/rnad.py).
I definitely plan to look more into this - if we want to make a presentation on RL for the others, I think it should be included.

**Reply by jbirkesteen** [reply]
I've found the article for free on [arxiv](https://arxiv.org/pdf/2206.15378.pdf) 😃

**Reply by jbirkesteen** [reply]
### Network architecture
I think we should do some more experiments with our architecture before implementing completely new methods. I know our observation space is much less complex than the one in Stratego, but still. Their architecture (and policy model, lol) is **insane** in comparison to ours 🤣  
This is of course related to the fact that they have history of past moves as part of their input (since it is an imperfect information game). But still, we might gain something by making slightly more complex architectures.
![image](https://user-images.githubusercontent.com/70960888/209484033-4b9ff909-7c1e-4c46-9580-bbe36eab72f4.png)
_(taken from the [deepmind article on deepnash for stratego](https://arxiv.org/pdf/2206.15378.pdf))_

**Comment by jbirkesteen** [comment]
For Go, deepmind used tree search methods. The observation space and state space in Connect Four is significantly smaller, so perhaps we could implement a (scaled down) similar approach?

**Comment by jbirkesteen** [comment]
I removed these links from `players.py` as I am currently trying to clean it out a bit, posting them here in case we might want to revisit them later.

> Links for implementing agents
https://medium.com/@carsten.friedrich/part-8-tic-tac-toe-with-policy-gradient-descent-da2496defc45
https://spinningup.openai.com/en/latest/spinningup/rl_intro3.html
http://rail.eecs.berkeley.edu/deeprlcourse-fa17/f17docs/lecture_4_policy_gradient.pdf

**Comment by jbirkesteen** [comment]
# Temporal difference methods
We have implemented `TDAgent` inspired by Sutton and Barto's walkthrough of Tesauro's TD-Gammon program in the RL book.
This is a short overview of our results so far.
## Tested training strategies

### Self-play
The tested version of self-play involves freezing newest parameters at the beginning of each episode. This seemingly results in a repetition of the same game every time, and of course very poor performance.

### Minimax
(CON-163)
Started out well, but had a long intermediary period of very bad performance. However, performance improved in the last part of the training and ended at approx. 27 % win-rate against Minimax with `max_depth=1`. Win-rate peaked at around 37 %.

Playing as a human against the agent highlights a fundamental flaw of this strategy: The agent has no idea that it can win in a "simple" way if its opponent makes a mistake.

### Half'n'half between self-play and Minimax
(CON-165)
This approach had one of the worst results yet, performing worse than 40 % win-rate against a random player.


### CON-170
self-play, freezes at beginning of each game (fixed)

weight values and value estimates are very high (after 10k games). 

What should our next steps be?

**Reply by jbirkesteen** [reply]
Change to epsilon-greedy action selection. Currently, we use a greedy strategy, always taking the action resulting in the best value estimate. It would be easy to change this approach into an epsilon-greedy method and use (some version of) self-play. This might avoid the problem of repetitious play.

**Reply by jbirkesteen** [reply]
Make the state representation more informative or the network architecture better at detecting patterns (or do some combination of both).

**Reply by jbirkesteen** [reply]
Change the opponent in some way. I fear that the minimax opponent might be too random - there is no plan in the way it is playing.

**Reply by jbirkesteen** [reply]
We could of course do some sort of hyperparameter search (in alpha, gamma or lambda), but I think my other suggestions have a better potential for bearing fruit.

**Reply by RasmusBrostroem** [reply]
Create a dataset of states and valuations that the agent can pre-train on in a supervised fashion. This way, the agent can get an idea of what it should do before trying to learn the game.

**Reply by jbirkesteen** [reply]
Try out other reward systems. Suggestions:
1. Give a negative reward at loss.
1. Give a reward each time the agent gets to place a piece, encouraging the agent to "survive" for longer. This reward should be dynamic (i.e. grow for each time step), but be relatively small compared to the reward for winning. 
1. Expand reward system, such that there are more rewards possible that guide the agents. 
    - If the agent has an opportunity to win, but does not take it then get negative reward
    - If the agent has opportunity to lose, but does not block it then get negative reward
